search/query/boolean.go

```diff
@@ -0,0 +1,101 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package query
+
+import (
+	"errors"
+
+	"github.com/m3db/m3ninx/index"
+	"github.com/m3db/m3ninx/search"
+	"github.com/m3db/m3ninx/search/searcher"
+)
+
+var (
+	errMustNotQuery = errors.New("queries which contain Must Not queries must contain at least one Must or Should query")
+)
+
+// booleanQuery is a compound query composed of must, should, and must not queries.
+type booleanQuery struct {
+	must    search.Query
+	should  search.Query
+	mustNot search.Query
+}
+
+// NewBooleanQuery returns a new compound query such that:
+//   - resulting documents match ALL of the must queries
+//   - resulting documents match AT LEAST ONE of the should queries
```

Comment:
i don't follow the rationale for `should`, do you have a use-case in mind for it?

Reply:
I'm gonna get rid of this. Boolean queries are common in general IR where documents can be ranked but make less sense for us where whether a document matches a query is a binary yes or no.